core/src/main/scala/pureconfig/ConfigConvert.scala

```diff
@@ -56,76 +56,139 @@ trait ConfigConvert[T] {
-    final def from(config: ConfigValue): Try[SubRepr] =
-      Failure(
-        new UnsupportedOperationException("Cannot call 'from' on a WrappedDefaultValueConfigConvert."))
-    def fromWithDefault(config: ConfigValue, default: DefaultRepr): Try[SubRepr] = config match {
-      case co: ConfigObject =>
-        fromConfigObject(co, default)
-      case other =>
-        Failure(WrongTypeException(config.valueType().toString))
+    final def from(config: ConfigValue): ConfigReaderResult[SubRepr] =
+      failWithThrowable(new IllegalStateException("Cannot call 'from' on a WrappedDefaultValueConfigConvert."))
+    def fromWithDefault(config: ConfigValue, default: DefaultRepr): ConfigReaderResult[SubRepr] = config match {
+      case co: ConfigObject => fromConfigObject(co, default)
+      case other => fail(WrongType(foundTyp = other.valueType().toString, expectedTyp = "ConfigObject"))
     }
-    def fromConfigObject(co: ConfigObject, default: DefaultRepr): Try[SubRepr]
+    def fromConfigObject(co: ConfigObject, default: DefaultRepr): ConfigReaderResult[SubRepr]
     def to(v: SubRepr): ConfigValue
   }
 
   implicit def hNilConfigConvert[Wrapped](
     implicit
     hint: ProductHint[Wrapped]): WrappedDefaultValueConfigConvert[Wrapped, HNil, HNil] = new WrappedDefaultValueConfigConvert[Wrapped, HNil, HNil] {
 
-    override def fromConfigObject(config: ConfigObject, default: HNil): Try[HNil] = {
-      if (!hint.allowUnknownKeys && !config.isEmpty) Failure(UnknownKeyException(config.keySet.iterator.next))
-      else Success(HNil)
+    override def fromConfigObject(config: ConfigObject, default: HNil): ConfigReaderResult[HNil] = {
+      if (!hint.allowUnknownKeys && !config.isEmpty) fail(UnknownKey(config.keySet.iterator.next))
+      else Right(HNil)
     }
 
     override def to(t: HNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
   }
 
-  private[pureconfig] def improveFailure[Z](result: Try[Z], keyStr: String): Try[Z] =
-    result recoverWith {
-      case CannotConvertNullException => Failure(KeyNotFoundException(keyStr))
-      case KeyNotFoundException(suffix) => Failure(KeyNotFoundException(keyStr + "." + suffix))
-      case WrongTypeException(typ) => Failure(WrongTypeForKeyException(typ, keyStr))
-      case WrongTypeForKeyException(typ, suffix) => Failure(WrongTypeForKeyException(typ, keyStr + "." + suffix))
+  private[pureconfig] def improveFailure[Z](failure: ConfigReaderFailure, keyStr: String): ConfigReaderFailure =
+    failure match {
+      case CannotConvertNull => KeyNotFound(keyStr)
+      case KeyNotFound(suffix) => KeyNotFound(keyStr + "." + suffix)
+      case WrongType(foundTyp, expectedTyp) => WrongTypeForKey(foundTyp, expectedTyp, keyStr)
+      case WrongTypeForKey(foundTyp, expectedTyp, suffix) => WrongTypeForKey(foundTyp, expectedTyp, keyStr + "." + suffix)
```

Comment:
Done with 0ab3792. With this PR I'm noticing how important is to model the errors that we can get and how those errors are propagated. I can see in future more PRs that could, for instance, give the context of an exception thrown.